Set: op= in-place, write some op in terms of op= for terseness

cassella opened the pull request:
Write the |=, +=, -=, and ^= to operate on lhs in-place rather than
using a temporary and replacing all of lhs with the temporary.  (The
temporary they were using was inside the |, +, -, or ^ operator they
invoked.)

This allows |= / += to not invalidate references to existing elements.
The other operations still do since they necessarily remove elements
from lhs.

Rewrite &= to be in-place when it's efficient to do so.  It still
invalidates references due to .remove().

Write | and ^ in terms of |= and ^= to reduce code duplication.

& and - weren't as amenable to writing in terms of &= and -=, for
efficiency reasons.

Since & is symmetric, and only has to iterate over one of its
arguments, choose to iterate over the smaller argument.

Add a test of each op= to the test of its op.

Correct name of module in module-ending comment.

dlongnecke-cray commented:
No, no particular reason. I can't recall my reasoning at the time at all but I obviously wasn't thinking about efficiency...

dlongnecke-cray commented:
@cassella I've bet you found that the calls to `|=` won't resolve properly? Just discovered this myself today...Goes to show what happens when you just assume something will work...

cassella commented:
Actually, I haven't touched this since I opened it, sorry.

The one I added in the test did resolve.  At least I'm 99% sure I ran that test before submitting.  :)

dlongnecke-cray commented:
No problem. Wait really? Now I'm even more confused...

cassella commented:
I rebased to master, and `test/library/standard/Set/` still passes, including the `|=` I added to `setUnion.chpl`.  What should I be being confused by?

dlongnecke-cray commented:
Nothing, it was on my end. Go ahead and ping me when you're ready for review.

cassella commented:
I had a further thought.  As it stands in this PR, the `|` and `|=` operators will contain a total of three loops over a set:

```chapel
  proc |(const ref a: set(?t, ?), const ref b: set(t, ?)): set(t) {
    var result: set(t, (a.parSafe || b.parSafe));

    for x in a do
      result.add(x);

    for x in b do
      result.add(x);

    return result;
  }

  proc |=(ref lhs: set(?t, ?), const ref rhs: set(t, ?)) {
    for x in rhs do
      lhs.add(x);
  }
```

I was thinking it would be more concise, though less natural, to turn it around and write `|` in terms of `|=` with just one loop:

```chapel
  proc |=(ref lhs: set(?t, ?), const ref rhs: set(t, ?)) {
    for x in rhs do
      lhs.add(x);
  }

  proc |(const ref a: set(?t, ?), const ref b: set(t, ?)): set(t) {
    var result: set(t, (a.parSafe || b.parSafe));

    result |= a;
    result |= b;

    return result;
}
```

I expect most if not all of the other operators to be amenable to the same sort of transformation.  What do you think of this approach?

dlongnecke-cray commented:
That seems reasonable to me.

cassella commented:
@dlongnecke-cray Ok.  The other operators didn't work out quite as well as `|` / `+` after all in terms of the `op=` not invalidating elements, since they all have the potential of removing elements from lhs.

dlongnecke-cray commented:
Bah, that's a good point! I will go ahead and take a look at this after I merge the collapse PR (I know, I know, I said I'd get to it Friday...)

dlongnecke-cray commented:
Ok, taking a look at this sometime today or tomorrow.

ronawho commented on modules/standard/Set.chpl:
```diff
@@ -627,7 +634,31 @@ module Set {
+          lhs.remove(x);
+      } else {
+        for x in rhs do
+          lhs.remove(x);
```
I mentioned this to @dlongnecke-cray already, but I think this branch is wrong -- it looks like its just implementing `-=` instead of `&=`

cassella replied:
Oops, thanks.  I opened  #15958 to cover that.
